Microsoft.HealthVault.Fhir/Transformers/LabTestResultsToFhir.cs

```diff
@@ -0,0 +1,250 @@
+
+            if (labTestResultGroup.LaboratoryName != null)
+            {
+                FhirOrganization fhirOrganization = labTestResultGroup.LaboratoryName.ToFhir();
+                observation.Contained.Add(fhirOrganization);
+                observation.Performer.Add(fhirOrganization.GetContainerReference());
+            }
+
+            observation.Status = GetStatus(labTestResultGroup.Status);
+
+            foreach (var subGroup in labTestResultGroup.SubGroups)
+            {
+                var subObservation = new Observation() { Id = "subObs" + Guid.NewGuid() };
+                ToFhirInternal(subGroup, subObservation);
+                observation.AddDomainResourceToContainer(subObservation);
+                observation.Related.Add(new Observation.RelatedComponent
+                {
+                    Type = Observation.ObservationRelationshipType.HasMember,
+                    Target = subObservation.GetContainerReference()
+                });
+            }
+
+            foreach (var resultDetail in labTestResultGroup.Results)
+            {
+                var resultObservation = new Observation() { Id = "resultObs" + Guid.NewGuid() };
+                ToFhirInternal(resultDetail, resultObservation);
+                observation.AddDomainResourceToContainer(resultObservation);
+                observation.Related.Add(new Observation.RelatedComponent
+                {
+                    Type = Observation.ObservationRelationshipType.DerivedFrom,
+                    Target = resultObservation.GetContainerReference()
+                });
+            }
+        }
+
+        private static void ToFhirInternal(LabTestResultDetails resultDetail, Observation resultObservation)
+        {
+            resultObservation.Effective = resultDetail.When?.ToFhir();
+
+            if (!string.IsNullOrEmpty(resultDetail.Name))
+            {
+                resultObservation.SetStringExtension(HealthVaultExtensions.LabTestResultName, resultDetail.Name);
+            }
+
+            if (resultDetail.Substance != null || resultDetail.CollectionMethod != null)
+            {
+                var specimen = new Specimen
+                {
+                    Id = "specimen" + Guid.NewGuid(),
+                    Type = resultDetail.Substance?.ToFhir(),
+                    Collection = resultDetail.CollectionMethod == null ? null : new Specimen.CollectionComponent
+                    {
+                        Method = resultDetail.CollectionMethod.ToFhir()
+                    }
+                };
+                resultObservation.AddDomainResourceToContainer(specimen);
+                resultObservation.Specimen = specimen.GetContainerReference();
+            }
+
+            if (resultDetail.ClinicalCode != null)
+            {
+                resultObservation.Code = resultDetail.ClinicalCode.ToFhir();
+            }
+
+            LabTestResultValue labTestResultValue = resultDetail.Value;
```

Comment:
Let's check the value first and then set the convenience variable inside the block.